jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/paths/PathCollection.java

```diff
@@ -0,0 +1,291 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.paths;
+
+import java.io.IOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.StringTokenizer;
+import java.util.function.BiPredicate;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import org.eclipse.jetty.util.StringUtil;
+
+/**
+ * Generic Loader for Multiple Path references.
+ * <p>
+ * This fits in the same space as the ResourceCollection of past Jetty versions.
+ * </p>
+ */
+public class PathCollection extends ArrayList<Path> implements AutoCloseable
```

Comment:
The `Path` class is immutable, so I think `PathCollection` should also be immutable as it will be often used as a `Path`.

So I think it should instead be an interface with static builder methods:
```java
interface PathCollection extends List<Path>,  AutoCloseable
{
    PathCollection from(Path... path);
    PathCollection from(List<Path> paths, Path... path);
    PathCollection from(List<Path> paths, Predicate<Path> filter);
    
```

Reply:
I agree that `PathCollection` should be immutable, which prevents it from extending from `ArrayList` or implementing `java.util.Collection` in any form.

Reply:
What about read-only apis for Stream or Iterable ?

Reply:
Do we need to be able to iterate the contained `Path`s, except for logging/monitoring purposes?

If you want to provide a powerful API to be able to select a specific `Path` using a complex query, you could eventually add a `stream()` method that would return a `Stream<Path>`.

Reply:
That shouldn't be a problem.  If the interface is a `List<String>`, then an implementation can use an immutable list behind that.

Reply:
I'm onboard with what @lorban is thinking.
I'm struggling to find a use case where iterating through the path entries has any purpose that isn't fulfilled by the API already.
Only one I can think of dump/debug.

Reply:
The only use-case that I can think of is that if `PathCollection` is immutable, than treating it as a `List<Path>` then it can be passed to a builder method like `PathCollection.from(List<Path>)` or better yet streamed/filtered or otherwise manipulated to create the a new `PathCollection`.

But it doesn't need to be the full `List` interface... perhaps just `Iterable`? maybe a `Stream<Path> stream()` method?

Eitherway, my main concern is that it is easy to create a new PathCollection from an old one. Something like:
```java
    PathCollection pathsBC = PathCollection.of(pathB, pathC);
    PathCollection pathsABC = PathCollection.of(pathA, pathsBC);
    PathCollection pathsABCDE = PathCollection.of(pathsABC, pathD, pathE);
```